### ISSUE TITLE
Parse Knex DSL like JSON objects

### ISSUE THREAD

**Issue by ayashjorden** [knex/knex#4010]
# Environment

Knex version: 0.21.2
Database + version: PG 11
OS: Linux (PhotosOS3

# Feature discussion / request

1. Explain what is your use case
I'd like to allow arbitrary select queries via GET API requests.

2. Explain what kind of feature would support this
Suggested is to add functionality to parse a DSL based JSON object:
3. Give some API proposal, how the feature should work
```typescript
await knex(TABLE_NAME).parse({
  "select" {
    "columns":[
      "colA",
      {name:"colB","operation":"count", "distinct":true, as:"aliasB"}
    ]
  }
})
```

@lorefnon , appreciate your feedback here as well.

Thank you,
Yarden

**Comment by elhigu** [comment]
I don't think knex should include any additional query syntaxes there are already too many of those. This could be implemented as a separate npm-package extension for knex.

**Comment by lorefnon** [comment]
Sorry for the delayed response: This feature does seem useful even if we don't add it to knex core.

@kibertoad Would knex-utils be a good place for a PR that implements this ?

**Comment by kibertoad** [comment]
@lorefnon yes, that sounds perfect

**Comment by ayashjorden** [comment]
Hi,
Appreciate your feedback.

How do we drive this forward? is there a weekly meeting to discuss such things?

**Comment by elhigu** [comment]
By adding an issue with the planned API to https://github.com/knex/knex-utils repository and then start trying out and drafting implementations + tests + documentation.

**Comment by intech** [comment]
Is there somewhere the possibility of converting from a query builder to a json schema inversely?

**Comment by intech** [comment]
@kibertoad @elhigu Is there somewhere the possibility of converting from a query builder to a json schema inversely?

I am looking for something similar to https://github.com/elastic/sql-summary, but this module is too simple and does not understand most of the complex syntax.
There is a constructor in knex that, with more advanced capabilities, can decompose a query into components and correctly group transactions, nested queries, constructs using `with` and ʻunion`.

**Comment by elhigu** [comment]
@intech no, knex query builder has pretty much zero support for introspection. To support that properly we would need to re-write it with some AST presentation of the query that should be built. Also knex doesn't have any SQL parsers (and it should never have one) for interpreting raw SQL.

**Comment by mercmobily** [comment]
Maintainers suggested implementing this in knex-utils so people who need a JSON DSL can opt-in without expanding core’s surface area. See the comments from @elhigu and @lorefnon pointing to knex-utils as the right home for this. That keeps knex lean and avoids locking us into yet another query syntax.

knex core’s builders are imperative; there is no AST that can be reliably serialised/deserialised (lib/query/querybuilder.js, lib/query/querycompiler.js). A robust DSL needs its own schema, validation and translation layer on top of the public builder API.

I recommend keeping this issue here, and then open a PR on knex-utils. Moving the issue to knex-util would turn it into abandonware almost certainly.  So, I am keeping this open as a feature request, but something that will land into knex-utils when done.
